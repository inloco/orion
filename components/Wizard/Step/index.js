import cx from 'classnames'
import PropTypes from 'prop-types'
import React, { Component } from 'react'

export class WizardStep extends Component {
  static propTypes = {
    className: PropTypes.string,
    children: PropTypes.node.isRequired
  }

  render() {
    const { children, className, ...otherProps } = this.props
    const classes = cx('inloco-wizard__step', className)
    return (
      <div className={classes} {...otherProps}>
        {children}
      </div>
    )
  }
}

export default WizardStep
