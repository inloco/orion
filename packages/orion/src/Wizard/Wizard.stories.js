import React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'
import { array, withKnobs } from '@storybook/addon-knobs'

import { Button, Wizard } from '../'

export const actions = {
  onStepIndexChange: action('onStepIndexChange'),
  onFinish: action('onFinish')
}

const buttons = {
  [Wizard.Buttons.NEXT]: <Button>Next</Button>,
  [Wizard.Buttons.BACK]: <Button>Return</Button>,
  [Wizard.Buttons.FINISH]: <Button onClick={actions.onFinish}>Complete</Button>
}

storiesOf('Wizard', module)
  .addDecorator(withKnobs)
  .add('basic', () => {
    return (
      <Wizard
        steps={array('Steps', ['Step 1', 'Step 2', 'Step 3'])}
        buttons={buttons}
        {...actions}>
        <div>
          Step 1 Content
          <div>
            Lorem, ipsum dolor sit amet consectetur adipisicing elit. Nemo
            veniam at, maxime adipisci sunt non voluptatum architecto quisquam
            sapiente eligendi natus nisi totam, incidunt, asperiores minus alias
            accusantium! Quis, aliquid.
          </div>
        </div>
        <div>
          Step 2 Content
          <div>
            Lorem ipsum dolor sit amet, consectetur adipisicing elit. Quia
            quidem quae ex deserunt commodi dolor vel rerum optio quos sunt
            quibusdam aspernatur sequi dolorum illum totam, non, dicta voluptate
            odit.
          </div>
        </div>
        <div>
          Step 3 Content
          <div>
            Lorem ipsum, dolor sit amet consectetur adipisicing elit. Eligendi
            laboriosam laudantium quam possimus esse, excepturi commodi ex
            cumque tenetur id ipsa in, est rerum sed fugit omnis incidunt
            provident! At!
          </div>
        </div>
      </Wizard>
    )
  })
