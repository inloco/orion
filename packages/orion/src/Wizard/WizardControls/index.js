import PropTypes from 'prop-types'
import React from 'react'

import Button from '../../Button'

export const WizardButtons = {
  BACK: 'back',
  SAVE: 'save',
  NEXT: 'next',
  FINISH: 'finish'
}

const WizardControls = ({
  buttons,
  currentStepIndex,
  onStepIndexChange,
  totalSteps
}) => {
  const isLastStep = currentStepIndex === totalSteps - 1
  return (
    <div className="wizard-controls">
      {currentStepIndex > 0 &&
        Button.create(buttons[WizardButtons.BACK], {
          autoGenerateKey: false,
          defaultProps: {
            onClick: () => onStepIndexChange(currentStepIndex - 1),
            type: 'button'
          }
        })}
      <div className="wizard-controls-right">
        {buttons[WizardButtons.SAVE] &&
          !isLastStep &&
          Button.create(buttons[WizardButtons.SAVE], {
            autoGenerateKey: false,
            defaultProps: {
              className: 'mr-8',
              ghost: true,
              type: 'submit'
            }
          })}
        {!isLastStep &&
          Button.create(buttons[WizardButtons.NEXT], {
            autoGenerateKey: false,
            defaultProps: {
              className: 'blue',
              onClick: () => onStepIndexChange(currentStepIndex + 1),
              primary: true,
              type: 'submit'
            }
          })}
        {isLastStep &&
          Button.create(buttons[WizardButtons.FINISH], {
            autoGenerateKey: false,
            defaultProps: {
              className: 'blue',
              primary: true,
              type: 'submit'
            }
          })}
      </div>
    </div>
  )
}

WizardControls.propTypes = {
  buttons: PropTypes.shape({
    [WizardButtons.BACK]: PropTypes.any,
    [WizardButtons.SAVE]: PropTypes.any,
    [WizardButtons.NEXT]: PropTypes.any,
    [WizardButtons.FINISH]: PropTypes.any
  }),
  currentStepIndex: PropTypes.number.isRequired,
  onStepIndexChange: PropTypes.func,
  totalSteps: PropTypes.number.isRequired
}

WizardControls.defaultProps = {
  buttons: {
    [WizardButtons.BACK]: 'Back',
    [WizardButtons.NEXT]: 'Next',
    [WizardButtons.FINISH]: 'Finish'
  },
  onStepIndexChange: () => {}
}

export default WizardControls
