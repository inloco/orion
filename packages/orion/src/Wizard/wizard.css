.orion.wizard .steps-nav {
  @apply justify-center;
}

.orion.wizard .wizard-controls {
  @apply flex mt-24;
}

.orion.wizard .wizard-controls-right {
  @apply ml-auto;
}
