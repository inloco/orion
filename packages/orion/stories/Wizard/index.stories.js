import React from 'react'
import { storiesOf } from '@storybook/react'

import { Placeholder, Wizard } from '../../components'

import './wizardStory.css'

storiesOf('Wizard', module).add('basic', () => (
  <Wizard steps={['First Step', 'Second Step', 'Third Step']}>
    <Wizard.Step>
      Step 1
      <Placeholder fluid>
        <Placeholder.Paragraph>
          <Placeholder.Line />
          <Placeholder.Line />
          <Placeholder.Line />
          <Placeholder.Line />
        </Placeholder.Paragraph>
        <Placeholder.Paragraph>
          <Placeholder.Line />
          <Placeholder.Line />
          <Placeholder.Line />
          <Placeholder.Line />
        </Placeholder.Paragraph>
        <Placeholder.Paragraph>
          <Placeholder.Line />
          <Placeholder.Line />
          <Placeholder.Line />
          <Placeholder.Line />
        </Placeholder.Paragraph>
        <Placeholder.Paragraph>
          <Placeholder.Line />
          <Placeholder.Line />
          <Placeholder.Line />
          <Placeholder.Line />
        </Placeholder.Paragraph>
        <Placeholder.Paragraph>
          <Placeholder.Line />
          <Placeholder.Line />
          <Placeholder.Line />
          <Placeholder.Line />
        </Placeholder.Paragraph>
        <Placeholder.Paragraph>
          <Placeholder.Line />
          <Placeholder.Line />
          <Placeholder.Line />
          <Placeholder.Line />
        </Placeholder.Paragraph>
        <Placeholder.Paragraph>
          <Placeholder.Line />
          <Placeholder.Line />
          <Placeholder.Line />
          <Placeholder.Line />
        </Placeholder.Paragraph>
      </Placeholder>
    </Wizard.Step>
    <Wizard.Step>Step 2</Wizard.Step>
    <Wizard.Step>Step 3</Wizard.Step>
  </Wizard>
))
