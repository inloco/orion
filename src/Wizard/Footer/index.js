import cx from 'classnames'
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import { Breadcrumb, Icon } from 'semantic-ui-react'

import WizardFooterBasicControls from './BasicControls'

class WizardFooter extends Component {
  static propTypes = {
    className: PropTypes.string,
    controls: PropTypes.func,
    currentStep: PropTypes.number,
    onChangeStep: PropTypes.func,
    steps: PropTypes.arrayOf(PropTypes.string).isRequired
  }

  static defaultProps = {
    currentStep: 0,
    onChangeStep: () => {}
  }

  render() {
    const {
      controls,
      className,
      currentStep,
      onChangeStep,
      steps,
      ...otherProps
    } = this.props
    const classes = cx('inloco-wizard__footer', className)
    return (
      <div className={classes} {...otherProps}>
        <Breadcrumb>
          {steps.map((step, index) => {
            const sectionClasses = cx({
              'inloco-wizard__footer-step--current': currentStep === index
            })
            return (
              <React.Fragment key={step}>
                {index !== 0 && (
                  <Breadcrumb.Divider>
                    <Icon className="chevron right" />
                  </Breadcrumb.Divider>
                )}
                <Breadcrumb.Section className={sectionClasses}>
                  {step}
                </Breadcrumb.Section>
              </React.Fragment>
            )
          })}
        </Breadcrumb>
        <div className="inloco-wizard__footer-controls">
          {this.renderControls({ currentStep, onChangeStep })}
        </div>
      </div>
    )
  }

  renderControls = props => {
    const { controls, steps } = this.props
    return controls ? (
      controls(props)
    ) : (
      <WizardFooterBasicControls {...props} totalSteps={steps.length} />
    )
  }
}

export default WizardFooter
