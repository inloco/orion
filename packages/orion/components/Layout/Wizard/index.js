import cx from 'classnames'
import PropTypes from 'prop-types'
import React, { Component } from 'react'

import { Layout } from '../'
import Wizard from '../../Wizard'

export class LayoutWizard extends Component {
  static propTypes = {
    className: PropTypes.string,
    children: PropTypes.node.isRequired,
    currentStep: PropTypes.number,
    headerExtra: PropTypes.node,
    initialStepIndex: PropTypes.number,
    mainContentProps: PropTypes.object,
    onStepChange: PropTypes.func,
    steps: PropTypes.arrayOf(PropTypes.string).isRequired
  }

  static defaultProps = {
    initialStepIndex: 0,
    mainContentProps: {}
  }

  state = {
    currentStep: this.props.initialStepIndex
  }

  isControlled = this.props.currentStep !== undefined

  render() {
    const {
      children,
      className,
      currentStep: currentStepProp,
      headerExtra,
      initialStepIndex,
      onStepChange,
      mainContentProps,
      steps,
      ...otherProps
    } = this.props
    const currentStep = this.isControlled
      ? currentStepProp
      : this.state.currentStep
    const classes = cx('inloco-layout__wizard', className)
    return (
      <React.Fragment>
        <Layout.Header>
          <Layout.HeaderTitle>
            <div className="inloco-layout__wizard-title">
              <div className="inloco-layout__wizard-title-number">
                {currentStep + 1}
              </div>
              {steps[currentStep]}
            </div>
          </Layout.HeaderTitle>
          {headerExtra && (
            <Layout.HeaderControls>{headerExtra}</Layout.HeaderControls>
          )}
        </Layout.Header>
        <Layout.MainContent {...mainContentProps}>
          <Wizard
            className={classes}
            currentStep={currentStep}
            steps={steps}
            onChangeStep={this.handleChangeStep}
            {...otherProps}>
            {children}
          </Wizard>
        </Layout.MainContent>
      </React.Fragment>
    )
  }

  handleChangeStep = newStep => {
    const { onStepChange } = this.props
    onStepChange && onStepChange(newStep)

    if (!this.isControlled) {
      this.setState({ currentStep: newStep })
    }
  }
}

export default LayoutWizard
