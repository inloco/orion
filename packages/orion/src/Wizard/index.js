import cx from 'classnames'
import _ from 'lodash'
import PropTypes from 'prop-types'
import React, { useState } from 'react'

import StepsNav from '../StepsNav'
import WizardControls, { WizardButtons } from './WizardControls'

const Wizard = ({
  children,
  className,
  currentStepIndex: currentStepProp,
  onStepIndexChange,
  steps,
  buttons
}) => {
  const [currentStepState, setCurrentStepState] = useState(currentStepProp || 0)
  const currentStepIndex = _.isNil(currentStepProp)
    ? currentStepState
    : currentStepProp

  const handleStepChange = newStep => {
    setCurrentStepState(newStep)
    onStepIndexChange(newStep)
  }

  return (
    <div className={cx('orion wizard', className)}>
      <StepsNav currentStep={currentStepIndex} steps={steps} />
      <div className="wizard-step">
        {React.Children.toArray(children)[currentStepIndex]}
      </div>

      <WizardControls
        buttons={buttons}
        currentStepIndex={currentStepIndex}
        onStepIndexChange={handleStepChange}
        totalSteps={steps.length}
      />
    </div>
  )
}

Wizard.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  currentStepIndex: PropTypes.number,
  onStepIndexChange: PropTypes.func,
  steps: PropTypes.arrayOf(PropTypes.string).isRequired,
  buttons: PropTypes.object
}

Wizard.defaultProps = {
  buttons: {
    [WizardButtons.BACK]: 'Back',
    [WizardButtons.NEXT]: 'Next',
    [WizardButtons.FINISH]: 'Finish'
  },
  onStepIndexChange: () => {}
}

Wizard.Buttons = WizardButtons

export default Wizard
