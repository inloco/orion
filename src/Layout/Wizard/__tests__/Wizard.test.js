import { shallow } from 'enzyme'
import React from 'react'

import LayoutWizard from '../'
import Wizard from '../../../Wizard'

const buildWrapper = props =>
  shallow(
    <LayoutWizard
      className="customTestClass"
      onStepChange={jest.fn()}
      steps={['Step 1', 'Step 2', 'Step 3']}
      {...props}>
      Test Child Content
    </LayoutWizard>
  )

describe('LayoutWizard', () => {
  it('should render layout wizard content for the first step by default', () => {
    expect(buildWrapper()).toMatchSnapshot()
  })

  it('should render layout wizard content for the given initial step', () => {
    const wrapper = buildWrapper({ initialStepIndex: 1 })
    expect(wrapper).toMatchSnapshot()
  })

  describe('when the "headerExtra" prop is given', () => {
    it('should should render the given node inside the header controls', () => {
      const wrapper = buildWrapper({ headerExtra: <div>Header Extra</div> })
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('when the current step changes', () => {
    let wrapper

    beforeEach(() => {
      wrapper = buildWrapper()
      const { onChangeStep } = wrapper.find(Wizard).props()
      onChangeStep(2)
    })

    it('should render the wizard at the new current step', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should call "onStepChange" callback with the new step', () => {
      const {
        props: { onStepChange }
      } = wrapper.instance()
      expect(onStepChange).toHaveBeenCalledWith(2)
    })
  })

  describe('when the component is "controlled"', () => {
    let wrapper

    beforeEach(() => {
      wrapper = buildWrapper({ currentStep: 1 })
    })

    it('should render the wizard at the given current step', () => {
      expect(wrapper).toMatchSnapshot()
    })

    describe('when the step changes in the <Wizard /> component', () => {
      beforeEach(() => {
        const { onChangeStep } = wrapper.find(Wizard).props()
        onChangeStep(2)
      })

      it('should keep rendering the wizard at the step specified by the "currentStep" prop', () => {
        expect(wrapper).toMatchSnapshot()
      })

      it('should call "onStepChange" callback with the new step', () => {
        const {
          props: { onStepChange }
        } = wrapper.instance()
        expect(onStepChange).toHaveBeenCalledWith(2)
      })
    })
  })
})
