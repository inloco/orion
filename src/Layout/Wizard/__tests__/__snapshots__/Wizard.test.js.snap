// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayoutWizard should render layout wizard content for the first step by default 1`] = `
<Fragment>
  <Header>
    <HeaderTitle>
      <div
        className="inloco-layout__wizard-title"
      >
        <div
          className="inloco-layout__wizard-title-number"
        >
          1
        </div>
        Step 1
      </div>
    </HeaderTitle>
  </Header>
  <MainContent>
    <Wizard
      className="inloco-layout__wizard customTestClass"
      currentStep={0}
      footerProps={Object {}}
      onChangeStep={[Function]}
      steps={
        Array [
          "Step 1",
          "Step 2",
          "Step 3",
        ]
      }
    >
      Test Child Content
    </Wizard>
  </MainContent>
</Fragment>
`;

exports[`LayoutWizard should render layout wizard content for the given initial step 1`] = `
<Fragment>
  <Header>
    <HeaderTitle>
      <div
        className="inloco-layout__wizard-title"
      >
        <div
          className="inloco-layout__wizard-title-number"
        >
          2
        </div>
        Step 2
      </div>
    </HeaderTitle>
  </Header>
  <MainContent>
    <Wizard
      className="inloco-layout__wizard customTestClass"
      currentStep={1}
      footerProps={Object {}}
      onChangeStep={[Function]}
      steps={
        Array [
          "Step 1",
          "Step 2",
          "Step 3",
        ]
      }
    >
      Test Child Content
    </Wizard>
  </MainContent>
</Fragment>
`;

exports[`LayoutWizard when the "headerExtra" prop is given should should render the given node inside the header controls 1`] = `
<Fragment>
  <Header>
    <HeaderTitle>
      <div
        className="inloco-layout__wizard-title"
      >
        <div
          className="inloco-layout__wizard-title-number"
        >
          1
        </div>
        Step 1
      </div>
    </HeaderTitle>
    <HeaderControls>
      <div>
        Header Extra
      </div>
    </HeaderControls>
  </Header>
  <MainContent>
    <Wizard
      className="inloco-layout__wizard customTestClass"
      currentStep={0}
      footerProps={Object {}}
      onChangeStep={[Function]}
      steps={
        Array [
          "Step 1",
          "Step 2",
          "Step 3",
        ]
      }
    >
      Test Child Content
    </Wizard>
  </MainContent>
</Fragment>
`;

exports[`LayoutWizard when the component is "controlled" should render the wizard at the given current step 1`] = `
<Fragment>
  <Header>
    <HeaderTitle>
      <div
        className="inloco-layout__wizard-title"
      >
        <div
          className="inloco-layout__wizard-title-number"
        >
          2
        </div>
        Step 2
      </div>
    </HeaderTitle>
  </Header>
  <MainContent>
    <Wizard
      className="inloco-layout__wizard customTestClass"
      currentStep={1}
      footerProps={Object {}}
      onChangeStep={[Function]}
      steps={
        Array [
          "Step 1",
          "Step 2",
          "Step 3",
        ]
      }
    >
      Test Child Content
    </Wizard>
  </MainContent>
</Fragment>
`;

exports[`LayoutWizard when the component is "controlled" when the step changes in the <Wizard /> component should keep rendering the wizard at the step specified by the "currentStep" prop 1`] = `
<Fragment>
  <Header>
    <HeaderTitle>
      <div
        className="inloco-layout__wizard-title"
      >
        <div
          className="inloco-layout__wizard-title-number"
        >
          2
        </div>
        Step 2
      </div>
    </HeaderTitle>
  </Header>
  <MainContent>
    <Wizard
      className="inloco-layout__wizard customTestClass"
      currentStep={1}
      footerProps={Object {}}
      onChangeStep={[Function]}
      steps={
        Array [
          "Step 1",
          "Step 2",
          "Step 3",
        ]
      }
    >
      Test Child Content
    </Wizard>
  </MainContent>
</Fragment>
`;

exports[`LayoutWizard when the current step changes should render the wizard at the new current step 1`] = `
<Fragment>
  <Header>
    <HeaderTitle>
      <div
        className="inloco-layout__wizard-title"
      >
        <div
          className="inloco-layout__wizard-title-number"
        >
          3
        </div>
        Step 3
      </div>
    </HeaderTitle>
  </Header>
  <MainContent>
    <Wizard
      className="inloco-layout__wizard customTestClass"
      currentStep={2}
      footerProps={Object {}}
      onChangeStep={[Function]}
      steps={
        Array [
          "Step 1",
          "Step 2",
          "Step 3",
        ]
      }
    >
      Test Child Content
    </Wizard>
  </MainContent>
</Fragment>
`;
